presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewQueryOptimizer.java

```diff
@@ -0,0 +1,267 @@
+import com.facebook.presto.sql.tree.LogicalBinaryExpression;
+import com.facebook.presto.sql.tree.Node;
+import com.facebook.presto.sql.tree.OrderBy;
+import com.facebook.presto.sql.tree.Query;
+import com.facebook.presto.sql.tree.QueryBody;
+import com.facebook.presto.sql.tree.QuerySpecification;
+import com.facebook.presto.sql.tree.Relation;
+import com.facebook.presto.sql.tree.Select;
+import com.facebook.presto.sql.tree.SelectItem;
+import com.facebook.presto.sql.tree.SimpleGroupBy;
+import com.facebook.presto.sql.tree.SingleColumn;
+import com.facebook.presto.sql.tree.SortItem;
+import com.facebook.presto.sql.tree.Table;
+import com.facebook.presto.sql.tree.TableSubquery;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+import java.util.Optional;
+
+import static java.util.Objects.requireNonNull;
+
+public class MaterializedViewQueryOptimizer
+        extends AstVisitor<Node, Void>
+{
+    private static final Logger logger = Logger.get(MaterializedViewQueryOptimizer.class);
+
+    private final Table materializedView;
+    private final ImmutableMap<String, String> baseToViewColumnMap;
+    private final Optional<Relation> baseQueryRelation;
+
+    public MaterializedViewQueryOptimizer(Table materializedView, Query originalSqlQuery)
+    {
+        this.materializedView = requireNonNull(materializedView, "materialized view is null");
+
+        QuerySpecification materializedViewDefinition = (QuerySpecification) originalSqlQuery.getQueryBody();
+        this.baseQueryRelation = materializedViewDefinition.getFrom().isPresent() ? materializedViewDefinition.getFrom() : Optional.empty();
+
+        // Create base to materialized view column map
+        ImmutableMap.Builder<String, String> baseToViewColumnMap = ImmutableMap.builder();
+
+        for (SelectItem baseTableSelectColumn : materializedViewDefinition.getSelect().getSelectItems()) {
+            if (baseTableSelectColumn instanceof SingleColumn) {
+                String baseColumnName = ((SingleColumn) baseTableSelectColumn).getExpression().toString();
+                Optional<Identifier> baseColumnAlias = ((SingleColumn) baseTableSelectColumn).getAlias();
+                String viewDerivedColumnName = baseColumnAlias.orElse(new Identifier(baseColumnName)).getValue();
+
+                baseToViewColumnMap.put(baseColumnName, viewDerivedColumnName);
+            }
+        }
+
+        this.baseToViewColumnMap = baseToViewColumnMap.build();
+    }
+
+    public Node rewrite(Node node)
+    {
+        try {
+            return process(node);
+        }
+        catch (Exception ex) {
+            logger.error(ex, "Failed to rewrite query: %s.", node);
+            return node;
+        }
+    }
+
+    @Override
+    public Node visitNode(Node node, Void context)
+    {
+        return node;
+    }
+
+    @Override
+    protected Node visitQuery(Query node, Void context)
+    {
+        return new Query(
```

Comment:
This class always creates new objects without checking if the rewrite is actually applied. Do we assume that this class is only used when we are sure the query can be rewritten by the materialization view? If that is the case, it is better to add a Class comment to indicate that there is another component to determine if a materialized view can be used to rewrite this SQL and this class is only used for eligible materialized views.

Reply:
@yuanzhanhku This work is divided in two parts. 
1. Validate if the query can be optimized, and if it can be, which MV will be chose for it.
2. Optimize the query.

This PR only handles with the #2. So it assumes that the given MV is good enough for rewrite. Hence it always rewrite. It is currently disabled by default for a configuration, which can be enabled for a query.

Reply:
Make sense. Thanks for the clarification.